File: python/src/aoc2025/secret_entrance.py
#!/usr/bin/env python3
#

class SecretEntrance:

  def __init__(self, turns=list()):
    self.turns = [[turn[0], int(turn[1:].strip())] for turn in turns]
    pass

  def zeroes(self):
    combo = 0
    position = 50

    for (direction, amount) in self.turns:
      if direction == 'L':
        position -= amount
      elif direction == 'R':
        position += amount

      position %= 100
      if not position:
        combo += 1

    return(combo)

  def clicks(self):
    combo = 0
    position = 50

    for (direction, amount) in self.turns:

      # every 100 def clicks
      combo += int(amount / 100)
      amount %= 100

      if direction == 'L':
        if (100 if position == 0 else position) <= amount:
          combo += 1
        position -= amount

      elif direction == 'R':
        if 100 <= (position + amount):
          combo += 1
        position += amount

      position %= 100

    return(combo)

File: ruby/lib/aoc2025/secret_entrance.rb
module Aoc2025
  class SecretEntrance
    attr_reader :turns

    def initialize turns=[]
      @turns = turns.inject([]) do |memo, item|
        memo.append [item[0], item[1..].to_i]
        memo
      end
    end

    def zeroes
      combo = 0

      @turns.inject(50) do |position, vector|
        direction, amount = vector
        if direction == 'L'
          position -= amount
        elsif direction == 'R'
          position += amount
        end

        position %= 100
        combo += 1 if position.zero?
        position
      end

      combo
    end

    def clicks
      combo = 0

      @turns.inject(50) do |position, vector|
        direction, amount = vector

        # every 100 def clicks
        combo += amount / 100
        amount %= 100

        # click if we scoot past or land on zero
        if direction == 'L'
          # amounts larger than the position click (use 100 for zero)
          combo += 1 if (position.zero? ? 100 : position) <= amount
          position -= amount
        elsif direction == 'R'
          # amounts that move the position past 100 click
          combo += 1 if 100 <= position + amount
          position += amount
        end

        position % 100
      end

      combo
    end

  end
end
